esp/esdllib/esdl_script.cpp

```diff
@@ -0,0 +1,914 @@
+            loadOtherwise(tree, prefix);
+        }
+    }
+
+    virtual ~CEsdlTransformOperationChoose(){}
+
+    bool process(IEspContext * context, IPropertyTree *request, IXpathContext * xpathContext) override
+    {
+        ForEachItemIn(i, m_conditionals)
+        {
+            if (m_conditionals.item(i).process(context, request, xpathContext))
+                return true;
+        }
+        return false;
+    }
+
+    virtual void toDBGLog () override
+    {
+    #if defined(_DEBUG)
+        DBGLOG (">>>>>>>>>>> %s >>>>>>>>>>", tagname.str());
+        ForEachItemIn(i, m_conditionals)
+            m_conditionals.item(i).toDBGLog();
+        DBGLOG (">>>>>>>>>>> %s >>>>>>>>>>", tagname.str());
+    #endif
+    }
+
+private:
+    void loadWhens(IPropertyTree * tree, const StringBuffer &prefix)
+    {
+        StringBuffer xpath;
+        Owned<IPropertyTreeIterator> children = tree->getElements(makeOperationTagName(xpath, prefix, "when"));
+        ForEach(*children)
+            m_conditionals.append(*new CEsdlTransformOperationConditional(&children->query(), prefix));
+    }
+
+    void loadOtherwise(IPropertyTree * tree, const StringBuffer &prefix)
+    {
+        StringBuffer xpath;
+        IPropertyTree * otherwise = tree->queryPropTree(makeOperationTagName(xpath, prefix, "otherwise"));
+        if (!otherwise)
+            return;
+        m_conditionals.append(*new CEsdlTransformOperationConditional(otherwise, prefix));
+    }
+};
+
+void processServiceAndMethodTransforms(std::initializer_list<IEsdlCustomTransform *> const &transforms, IEspContext * context, IPropertyTree *tgtcfg, const char *service, const char *method, const char* reqtype, StringBuffer & request, IPropertyTree * bindingCfg)
+{
+    LogLevel level = LogMin;
+    if (!transforms.size())
+        return;
+    if (tgtcfg)
+        level = (unsigned) tgtcfg->getPropInt("@traceLevel", level);
+
+    if (request.length()!=0)
+    {
+        if (level >= LogMax)
+        {
+            DBGLOG("ORIGINAL REQUEST: %s", request.str());
+            StringBuffer marshalled;
+            if (bindingCfg)
+                toXML(bindingCfg, marshalled.clear());
+            DBGLOG("BINDING CONFIG: %s", marshalled.str());
+            if (tgtcfg)
+                toXML(tgtcfg, marshalled.clear());
+            DBGLOG("TARGET CONFIG: %s", marshalled.str());
+        }
+
+        bool strictParams = bindingCfg ? bindingCfg->getPropBool("@strictParams", false) : false;
+        Owned<IXpathContext> xpathContext = getXpathContext(request.str(), strictParams);
+
+        StringArray prefixes;
+        for ( IEsdlCustomTransform * const & item : transforms)
+        {
+            if (item)
+                item->appendEsdlURIPrefixes(prefixes);
+        }
+
+        registerEsdlXPathExtensions(xpathContext, context, prefixes);
+
+        VStringBuffer ver("%g", context->getClientVersion());
+        if(!xpathContext->addInputValue("clientversion", ver.str()))
```

Comment:
I believe this is intended to be a call to addVariable so that an explicit param declaration is **not** required when strictParams is enabled. That's because clientversion is used so frequently (along with query, method, service and request).

Reply:
agreed.